Improve weekly bug-checks

cpaelzer opened the pull request:
One can save the reported list of tagged bugs like
```
$ ustriage --no-show-triage --extended --show-tagged --tag server-next --flag-recent 6 -S ~/savebugs/next-$(date -I'seconds').yaml
...
Saved reported bugs in /home/paelzer/savebugs/next-2022-02-01T12:45:10+01:00.yaml
```

And the week after (or any time later) one can refer to the old list:
```
$ ustriage --no-show-triage --extended --show-tagged --tag server-next --flag-recent 6 -S ~/savebugs/next-$(date -I'seconds').yaml -C ~/savebugs/next-2022-02-01T12\:45\:10+01\:00.yaml
```

This will mark bugs not yet in last weeks list as "N" for new like
```
LP: #1951490 -  +UN Confirmed     [samba]             27.01.22 High                     - Can't print after update to 4.13
```

And it will add an extra list of bugs that are gone (closed, dropped tag, ...)
```
Bugs gone compared with /home/paelzer/savebugs/next-2022-02-01T12:45:10+01:00.yaml:
Found 3 bugs
LP: #1890230 -   UN Confirmed     [ipxe]              27.01.22 Medium                   - arm64 ipxe package isn't an arm64 build
also:        -   U  Confirmed     [ipxe]                       Medium
also:        -   U  Won't Fix     [ipxe]                       Medium
```

_P.S. While working on this I also improved the flags readability by making each use fix-columns._

paride commented on ustriage/ustriage.py:
```diff
@@ -763,6 +808,14 @@ def launch():
+    parser.add_argument('-C', '--compare-tagged-bugs-to',
+                        default=False,
+                        dest='filename_compare',
+                        help='Compare the reported tagged bugs to %file')
```
The `%` in `%file` break `--help`, the error is `TypeError: must be real number, not dict`, but I think that's a bit bogus and the real reason is that the `%` is interpreted as a format string. I removed the `%` to verify that's the culprit and it works just fine.

cpaelzer replied:
Thanks, fixed now

paride reviewed (changes requested):
One inline comment, for the rest works fine and LGTM.

Non-blocking comment:

As we're now dumping data to a file, we kind of implicitly commit to a file format. I don't think we want to do that, to be free to update (= break) it in any moment. It may be worth adding a note in the `README.md` (something like: *[...] for the moment the SAVE/COMPARE file format has no guarantee of stability across releases*).

bryceharrington commented on ustriage/task.py:
```diff
@@ -136,12 +136,12 @@ def short_title(self):
+        if (self.AGE and self.date_last_updated > self.AGE):
             flags += 'U'
+        else:
+            flags += ' '
```
For consistency might be better to stick with the same if else block style:

`flags += 'U' if (self.AGE and self.date_last_updated > self.AGE) else ' '
`

cpaelzer replied:
I had this initially, it breaks the >79 rule and thereby tox and our CI - that is the reason I had to fall back to this.
I could only use '\' as a continuation line, which I thought to be worse than the full if as it starts a new style (continuation lines) which we didn't have so far.

bryceharrington commented on ustriage/task.py:
```diff
@@ -190,7 +191,7 @@ def compose_dup(self, shortlinks=True, extended=False):
         flags = self.get_flags()
 
-        text = '%s - %3s %-13s %-19s' % (
+        text = '%s - %4s %-13s %-19s' % (
```
Since get_flags() now returns a fixed length string, it's unnecessary to specify the field width here, and will be more maintainable if it doesn't.  I.e., you just need:

`         text = '%s - %s %-13s %-19s' % (`

both here and above.

cpaelzer replied:
Agreed, and I updated the docstring of the `get_flags` to outline that it has to be fixed-length.

bryceharrington commented on ustriage/task.py:
```diff
@@ -161,9 +162,9 @@ def compose_pretty(self, shortlinks=True, extended=False):
-        text = '%s - %3s %-13s %-19s' % (
+        text = '%s - %4s %-13s %-19s' % (
             bug_url,
             flags,
```
The flags variable isn't used any other places, so maybe consider just inlining the get_flags() call here and eliminating the variable?

cpaelzer replied:
Agreed, IIRC this come from an older control flow in the code.
Fixed - thanks for the suggestion.

bryceharrington commented on ustriage/ustriage.py:
```diff
@@ -272,7 +273,7 @@ def parse_dates(start, end=None):
 def print_bugs(tasks, open_in_browser=False, shortlinks=True, blacklist=None,
                limit_subscribed=None, oder_by_date=False, is_sorted=False,
-               extended=False):
+               extended=False, filename_save=False, filename_compare=False):
```
The default (unset) filenames should be =None rather than =False for proper type consistency.

cpaelzer replied:
It is Default False and thereby a bool everywhere (consistently), not just here - also in all function definitions, argparse and so on. And a type string if set via the commandline.

Do you mean with type consistency that, since the set value isn't `True` then the unset shouldn't be `False`.
And therefore instead it should be `None` or `String`?
Ok - under that aspect that this is closer I can follow the argument.

Fixed, thanks for the hint!

bryceharrington replied:
Yep, that's it precisely.

bryceharrington reviewed (changes requested):
A few stylistic fixups listed inline.

I like this approach.  When I suggested the idea I had imagined using some Launchpad API logic to extract bugs closed_since '-1 week', or something, but that might be an expensive (i.e. time consuming) calculation.  This approach takes advantage of the fact that if the person cares, it's because they are probably the one that generated the report previously so can be expected to have saved the data.  Also cron-friendly.  ;-)

For the saved data, personally I prefer JSON for script-to-script data sharing, and YAML just for human-to-script data sharing (e.g. config files).  But for this case it doesn't make much difference.

cpaelzer commented:
> One inline comment, for the rest works fine and LGTM.
> 
> Non-blocking comment:
> 
> As we're now dumping data to a file, we kind of implicitly commit to a file format. I don't think we want to do that, to be free to update (= break) it in any moment. It may be worth adding a note in the `README.md` (something like: _[...] for the moment the SAVE/COMPARE file format has no guarantee of stability across releases_).

Thanks for he suggestion.
I have updated the readme for this and some other yet undocumented housekeeping use cases.
Also fixed a few typos aspell was showing.

cpaelzer commented:
I think I have fixed or explained all outstanding feedback - new content pushed and asking for re-review please.

paride commented:
> For the saved data, personally I prefer JSON for script-to-script data sharing, and YAML just for human-to-script data sharing (e.g. config files). But for this case it doesn't make much difference.

This is what I was thinking when suggesting to specify that we don't commit to a file format :) But indeed no much difference in this case.

paride reviewed (approved):
LGTM, thanks! Leaving the merge to you so you can chose your preferred merge style or wait or wait for more feedback from Bryce.

utkarsh2102 commented on README.md:
```diff
@@ -98,7 +98,7 @@ Those are "date of the last update", "importance" and "assignee" (if there is an
 
 The expiration is defined as 60 days of inactivity in server-next tagged bugs, and 180 days for the other ubuntu-server subscribed bugs.
-These durations as well as the tag it considers for the "active" list can be tuned via the arguments, --expire-tagged, --expire and --tag.
+These duration's as well as the tag it considers for the "active" list can be tuned via the arguments, --expire-tagged, --expire and --tag.
```
Um, shouldn't it be `durations` only? :o

cpaelzer replied:
Well aspell suggested that and I didn't think too deeply into it.
But you are right it isn't showing ownership (genitive) for sure so yes this is wrong.
/me stares at aspell and re-fixes it :-)

utkarsh2102 commented on README.md:
```diff
@@ -116,8 +116,29 @@ ubuntu-server as our current two levels of [bug tracking](https://github.com/can
+meeting. Since the situation in launchpad might have changed (dropped tag,
+closed the bug, assigned to other teams, changed subscription) and not all of
+them can be detected from launchpad-api after the fast ustriage now also
+provides the option to save and compare a list of stored bugs
```
```suggestion
provides the option to save and compare a list of stored bugs.
```

cpaelzer replied:
agreed.

utkarsh2102 reviewed (approved):
Minor two fixes, rest looks good, I believe.

cpaelzer commented:
@bryceharrington are all your questions/suggestions resolved so that we can merge this?

bryceharrington commented:
Yep, looks good!
